kernelci.org/content/en/docs/tests/howto.md

```diff
@@ -1,334 +1,336 @@
 
-Genrate test job definitions
+and to test it would fail if the OS name was not the expected one:
 
 ```
-kci_test generate --bmeta-json /path/to/bmeta.json --dtbs-json /path/to/artifacts/dtbs.json --lab-json /path/to/mgalka-lava-local.json --storage http://storage.kernelci --lab-config your-lab-name --user lava_user --lab-token LAVA_API_TOKEN --output /output/path --callback-id your-lava-callback-name --callback-url http://api.kernelci
+$ ./uname-os OpenBSD
+System: 'Linux', expected: 'OpenBSD', result: FAIL
+$ echo $?
+1
 ```
 
-**Note**
-
-Make sure that there is that the lab you chose to supports the type of tests which you've just defined. You may need to modify `lab-configs.yaml` accordingly.
+Now, let's see how this can be used with KernelCI.
 
-e.g.
+### Build it and run the C implementation
 
-```yaml
-  your-local-lab:
-    lab_type: lava
-    url: 'http://kernelci-lava/RPC2/'
-    filters:
-      - passlist:
-          plan:
-            - baseline
-            - btrfs
-```
+See commit on the [how-to
+branch](https://github.com/kernelci/kernelci-core/commits/how-to):
+`config/lava: download and build uname-os.c and use it`
 
+Any arbitrary commands can be added to the `uname.jinja2` template before
+running the actual test cases.  In this example, we can install Debian packages
+then download the `uname-os.c` file and compile it to be able to finally run it
+as a test case:
 
-Submit test jobs
+```yaml
+          steps:
+          - apt update
+          - apt install -y wget gcc
+          - wget https://raw.githubusercontent.com/kernelci/kernelci-core/how-to/config/lava/uname/uname-os.c
+          - gcc -o uname-os uname-os.c
+          - lava-test-case uname-os-shell --shell '[ $(uname -o) = "GNU/Linux" ]'
+          - lava-test-case uname-os-c --shell './uname-os'
+```
+
+Please note that if one of the steps fails, the job will abort.  So if `apt
+install` or `wget` fails, the tests won't be run and the LAVA job status will
+show an error.
+
+We now have 2 test cases, one with the shell version and one with the C
+version.  After updating the pull request on GitHub, this will also get tested
+automatically on staging.  Here's some sample results:
+
+[Results for uname: «staging-mainline-20210805.0» on «minnowboard-max-E3825» (kernelci / staging-mainline)](https://staging.kernelci.org/test/plan/id/610bc171b2d76ba86d2a361e/)
+
+## Step 3: Going further
+
+With Step 2, pretty much anything can already be run within the limitations of
+the CPU and network bandwidth on the target platform.  Even if this doesn't
+take too long to run, there are many reasons why it's not really suitable to
+enable in production.  There's no point installing the same packages and
+building the same source code over and over again, it will add up as
+significant wasted resources and extra causes of test job failures.  Once the
+steps require to run a test suite are well defined, having a rootfs image with
+everything pre-installed solves these issues.
```

Comment:
Perhaps it's also worth noting somewhere that not every LAVA lab allows direct internet access to the hosted platforms, so relying on this approach isn't 100% portable.

Reply:
That's true, I'll add something about it.  Also spotted a typo "once the steps require*d*" - I'll fix that at the same time.

Reply:
Also added a note block about that.

Reply:
LGTM